File: auto_data_collection.py
import os
import math
import zipfile
import pickle
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore", category=UserWarning, module="openpyxl")

# ------------
# unzip file
# ------------
date = 20260107
file_name = r"D:\研究\weekly"

zipfile_path = os.path.join(os.path.expanduser("~"), "Downloads", f"{date}_cs_newexp.zip")
with zipfile.ZipFile(zipfile_path, "r") as zf:
    zf.extractall(file_name)

data_dir = os.path.join(file_name, f"{date}_cs_newexp")
if not os.path.isdir(data_dir):
    cand = [
        os.path.join(file_name, d)
        for d in os.listdir(file_name)
        if os.path.isdir(os.path.join(file_name, d))
        and str(date) in d
        and "cs_newexp" in d
    ]
    if cand:
        data_dir = cand[0]
    else:
        raise FileNotFoundError(f"Cannot find extracted folder like {date}_cs_newexp under {file_name}")

# ------------
# load data
# ------------
batt_cycle_df = {}  
statis_df     = {}   
capacity_map  = {}   
cycle_map     = {}   

def loadfunction(i: int, j: int):
    cands = range(59, 64)  # 59:63
    fns = [os.path.join(data_dir, f"240076-{i}-{j}-28185739{c}.xlsx") for c in cands]
    existing = [fn for fn in fns if os.path.exists(fn)]
    if not existing:
        raise FileNotFoundError("Cannot find any file among: " + ", ".join(fns))
    N = existing[0]

    df_step  = pd.read_excel(N, sheet_name="step", engine="openpyxl")
    df_cycle = pd.read_excel(N, sheet_name="cycle", engine="openpyxl")

    cp_idx = df_step["工步類型"].astype(str).eq("恆流放電")
    cap = df_step.loc[cp_idx, "容量(Ah)"].dropna().to_numpy()

    k = 10 * i + j
    batt_cycle_df[k] = df_cycle
    statis_df[k]     = df_step
    capacity_map[k]  = cap
    cycle_map[k]     = int(len(cap))

for i in range(1, 5):
    for j in range(1, 9):
        loadfunction(i, j)
        print(f"Battery {i}-{j} Done")

# -------------------
# generate report
# -------------------
report_dir = os.path.join(data_dir, "report")
os.makedirs(report_dir, exist_ok=True)
os.chdir(report_dir)

i_list = list(range(1, 5))
ja = list(range(1, 9, 2))  # 1,3,5,7
jb = list(range(2, 9, 2))  # 2,4,6,8

def build_plan_indices(i_list, j_list):
    idx_mat = np.array([[10 * ii + jj for ii in i_list] for jj in j_list])
    idx_vec = idx_mat.flatten(order="F")
    return idx_vec

idxA = build_plan_indices(i_list, ja)
idxB = build_plan_indices(i_list, jb)

PlanA_cycle = [cycle_map[k] for k in idxA]
PlanB_cycle = [cycle_map[k] for k in idxB]

PlanA_capacity = [capacity_map[k] for k in idxA]
PlanB_capacity = [capacity_map[k] for k in idxB]

PlanA_standcapacity = [cap / cap[0] for cap in PlanA_capacity]
PlanB_standcapacity = [cap / cap[0] for cap in PlanB_capacity]

battery_name_planA = [f"Battery {k//10}-{k%10}" for k in idxA]
battery_name_planB = [f"Battery {k//10}-{k%10}" for k in idxB]

PlanA_dat_plot = dict(standcapacity=PlanA_standcapacity, cycle=PlanA_cycle, Battery_name=battery_name_planA)
PlanB_dat_plot = dict(standcapacity=PlanB_standcapacity, cycle=PlanB_cycle, Battery_name=battery_name_planB)

xlim_num = [min(PlanA_cycle), min(PlanB_cycle)]

def overall_ylim(standcaps, xlim):
    mins, maxs = [], []
    for arr in standcaps:
        seg = arr[:xlim]
        if len(seg) == 0:
            continue
        mins.append(float(np.min(seg)))
        maxs.append(float(np.max(seg)))
    return (min(mins) - 0.01, max(maxs) + 0.01)

cap_range = [overall_ylim(PlanA_standcapacity, xlim_num[0]),
             overall_ylim(PlanB_standcapacity, xlim_num[1])]

state = {
    "date": date,
    "data_dir": data_dir,
    "report_dir": report_dir,
    "batt_cycle_df": batt_cycle_df,
    "statis_df": statis_df,
    "capacity_map": capacity_map,
    "cycle_map": cycle_map,
    "PlanA_dat_plot": PlanA_dat_plot,
    "PlanB_dat_plot": PlanB_dat_plot,
    "xlim_num": xlim_num,
    "cap_range": cap_range,
    "idxA": idxA,
    "idxB": idxB,
}
with open(f"{date}_cs_newexp.pkl", "wb") as f:
    pickle.dump(state, f)

def make_col_pattern(xlim, mode):
    if mode == "each":
        base = np.repeat([0, 1, 2], 20)
    elif mode == "times":
        base = np.tile([0, 1, 2], 20)
    else:
        raise ValueError("mode must be 'each' or 'times'")
    rep = int(math.ceil(xlim / len(base)))
    return np.tile(base, rep)[:xlim]

def plot_plan(out_png, plan, xlim, ylim, vline, col_mode):
    fig, axes = plt.subplots(4, 4, figsize=(12, 12), dpi=600)
    axes = axes.flatten()

    col_vals = make_col_pattern(xlim, col_mode)

    for idx in range(16):
        ax = axes[idx]
        y = plan["standcapacity"][idx][:xlim]
        x = np.arange(1, len(y) + 1)

        ax.scatter(x, y, c=col_vals[:len(y)], cmap="tab10", s=14)
        ax.set_xlim(0, xlim)
        ax.set_ylim(ylim[0], ylim[1])
        ax.set_xlabel("cycle", fontsize=14)
        ax.set_ylabel("Standard Capacity", fontsize=14)
        ax.set_title(plan["Battery_name"][idx], fontsize=14)
        ax.axvline(vline, color="gray", linewidth=1)
        ax.tick_params(axis="both", labelsize=12)

    plt.tight_layout()
    fig.savefig(out_png)
    plt.close(fig)

plot_plan("PlanA_capacity.png", PlanA_dat_plot, xlim_num[0], cap_range[0], vline=28, col_mode="each")
plot_plan("PlanB_capacity.png", PlanB_dat_plot, xlim_num[1], cap_range[1], vline=31, col_mode="times")

File: auto_data_collection.R
##############
# unzip file #
##############
library(readxl)
library(tidyverse)
library(lubridate)

date <- 20260107
file_name <- "D:\\研究\\weekly"

zipfile <- file.path(Sys.getenv("USERPROFILE"), 
                     "Downloads", 
                     paste0(date,"_cs_newexp.zip"))
unzip(zipfile, exdir = file_name)

#############
# load data #
#############
setwd(file.path(file_name,paste0(date,"_cs_newexp")))
loadfunction <- function(i,j){
  
  cands <- 59:63
  fns <- paste0("240076-", i, "-", j, "-28185739",cands,".xlsx")
  idx <- which(file.exists(fns))[1]
  if (is.na(idx)) stop("Cannot find any file among：", paste(fns, collapse = ", "))
  N <- fns[idx]
  
  TS <- "step"
  part.T <- as.data.frame(read_excel(N, sheet = TS))
  
  S <- "cycle"
  part.NS <- as.data.frame(read_excel(N, sheet = S))
  
  cp_idx <- part.T$工步類型 == "恆流放電"
  capacity <- part.T$`容量(Ah)`[cp_idx]
  cycle <- length(capacity)
  
  name <- paste0("battery",i,j)
  assign(name, part.NS, envir = parent.frame())
  name2 <- paste0("capacity",i,j)
  assign(name2, capacity, envir = parent.frame())
  nameC <- paste0("cycle",i,j)
  assign(nameC, cycle, envir = parent.frame())
  
  statis <- paste0("statis",i,j)
  assign(statis, part.T, envir = parent.frame())
}


for(i in 1:4){
  for(j in 1:8){
    loadfunction(i,j) 
    cat("Battery",i,"-",j," Done\n")
  }
}


###################
# generate report #
###################
dir.create(file.path(file_name,paste0(date,"_cs_newexp"),"report"), 
           showWarnings = FALSE, recursive = TRUE) 
setwd(file.path(file_name,paste0(date,"_cs_newexp"),"report")) 

i <- 1:4
ja <- seq(1,8,by=2)
jb <- seq(2,8,by=2)

# plan A
idx <- t(outer(i, ja, function(ii, jj) 10*ii + jj))
PlanA_cycle <- sapply(idx, function(k) {
  x <- get(paste0("cycle", k), envir = .GlobalEnv)
  x
})
PlanA_capacity <- sapply(idx, function(k) {
  x <- get(paste0("capacity", k), envir = .GlobalEnv)
  x
})
PlanA_standcapacity <- sapply(1:16,function(j){
  PlanA_capacity[[j]]/PlanA_capacity[[j]][1]
})
battery_name_planA <- paste0("Battery ",idx%/%10,"-",idx%%10)

# plan B
idx <- t(outer(i, jb, function(ii, jj) 10*ii + jj))
PlanB_cycle <- sapply(idx, function(k) {
  x <- get(paste0("cycle", k), envir = .GlobalEnv)
  x
})
PlanB_capacity <- sapply(idx, function(k) {
  x <- get(paste0("capacity", k), envir = .GlobalEnv)
  x
})
PlanB_standcapacity <- sapply(1:16,function(j){
  PlanB_capacity[[j]]/PlanB_capacity[[j]][1]
})
battery_name_planB <- paste0("Battery ",idx%/%10,"-",idx%%10)


PlanA_dat_plot <- list(standcapacity = PlanA_standcapacity,
                       cycle = PlanA_cycle, 
                       Battery_name = battery_name_planA)
PlanB_dat_plot <- list(standcapacity = PlanB_standcapacity,
                       cycle = PlanB_cycle, 
                       Battery_name = battery_name_planB)

xlim_num <- c(min(PlanA_cycle),
              min(PlanB_cycle))
cap_range <- list(range(sapply(PlanA_dat_plot$standcapacity, function(x) range(x[1:xlim_num[1]])))+c(-1,1)*0.01,
                  range(sapply(PlanB_dat_plot$standcapacity, function(x) range(x[1:xlim_num[2]])))+c(-1,1)*0.01)

save.image(file = paste0(date,"_cs_newexp.Rdata"))

png("PlanA_capacity.png", 
    width = 8*1.5, height = 8*1.5, units = "in",res = 600)
col_ss = rep(1:3,each = 20)
par(mfrow = c(4,4))
for(i in 1:16){
  plot(1:xlim_num[1], PlanA_dat_plot$standcapacity[[i]][1:xlim_num[1]], col = col_ss, ylim = cap_range[[1]],xlim = c(0,xlim_num[1]),
       pch = 16, xlab = "cycle", ylab = "Standard Capacity", main = PlanA_dat_plot$Battery_name[i],
       cex.axis = 1.8, cex.lab = 1.8, cex.main = 1.8,cex=1.5)
  abline(v=28,col="gray")
}
dev.off()


png("PlanB_capacity.png", 
    width = 8*1.5, height = 8*1.5, units = "in",res = 600)
col_ss = rep(1:3,times = 20)
par(mfrow = c(4,4))
for(i in 1:16){
  plot(1:xlim_num[2], PlanB_dat_plot$standcapacity[[i]][1:xlim_num[2]], col = col_ss, ylim = cap_range[[2]],xlim = c(0,xlim_num[2]),
       pch = 16, xlab = "cycle", ylab = "Standard Capacity", main = PlanB_dat_plot$Battery_name[i],
       cex.axis = 1.8, cex.lab = 1.8, cex.main = 1.8,cex=1.5)
  abline(v=31,col="gray")
}
dev.off()


